Add cold/offline backup script

fvaleri opened the pull request:
### Type of change
- Enhancement / new feature

### Description
Bash script for cold/offline backups of Kafka clusters on Kubernetes/OpenShift.

If you think you do not need a backup strategy for Kafka because of its embedded data replication,
then consider the impact of a misconfiguration, bug or security breach that deleted all your data.
For hot/online backups, you can use storage snapshotting or streaming into object storage.

To run the script, the Kubernetes user must have permission to work with PVC and Strimzi custom resources.
The procedure will stop the Cluster Operator and selected cluster for the duration of the backup. Before
restoring the Kafka cluster you need to make sure to have the right version of Strimzi CRDs installed.
If you have a single cluster-wide Cluster Operator, then you need to scale it down manually. You can run
backup and restore procedures for different Kafka clusters in parallel. Only the local file system is supported.
Consumer group offsets are included, but not Kafka Connect, MirrorMaker and Kafka Bridge custom resources.

### Requirements
- bash 5+ (GNU
- tar 1.33+ (GNU)
- kubectl 1.16+ (K8s CLI)
- yq 4.6+ (YAML processor)
- zip 3+ (Info-ZIP)
- unzip 6+ (Info-ZIP)
- enough disk space

scholzj commented:
TBH I'm not sure it belongs here. Maybe it should live in its own repository? We have already now problems with how long the build takes. Adding another container image will make it even slower. So we should try to peel things off instead of adding more - and this does not seem to have any dependencies on any other parts of this repo, so it should be easy to have it in separate repository.

Also, can you maybe explain a bit more in detail how does it work? I think that would be useful also as a general information in the readme to have anyone using it understand how it works and what will it do.

scholzj commented on docker-images/cold-backup/Dockerfile:
```diff
@@ -0,0 +1,3 @@
+FROM alpine:3.13
```
This should use the standard base image - either the strimzi base image or CentOS 7.

fvaleri replied:
No problem with this, it just seems an overkill for `rsync` only. Given that you are doing a backup, the base image is already cached, nevertheless Alpine based images are lighter and, most importantly, quicker at runtime.

scholzj commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,252 @@
+    done
+
+    # stop operator and statefulsets
+    oc scale deployment strimzi-cluster-operator --replicas 0
```
Scaling down the operator is too disruptive since it affects also other clusters. Thsi should make use of the Pause feature which is being worked on in https://github.com/strimzi/strimzi-kafka-operator/pull/4399

fvaleri replied:
That's interesting, thanks.

scholzj commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,252 @@
+    # stop operator and statefulsets
+    oc scale deployment strimzi-cluster-operator --replicas 0
+    oc scale statefulset $CLUSTER_NAME-kafka --replicas 0
+    oc scale statefulset $CLUSTER_NAME-zookeeper --replicas 0
```
Do we need to scaledown all components? Entity Operator, Cruise Control, Kafka Exporter? They will be just looping and generating errors while the KAfka / Zoo are not available and possibly trigger alerts.

fvaleri replied:
Not sure about this. Alerts will fire also if you shutdown components.

scholzj commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,252 @@
+
+restore() {
+    __confirm $TARGET_NS
+    oc new-project $TARGET_NS 2>/dev/null || oc project $TARGET_NS
```
This should use `kubectl`.

fvaleri replied:
I considered that. The only problem I have in switching to `kubectl` is that the `rsync` flag is not available there and this is a foundamental part of the script. I will see if it's possible to replace it with a new function (like I did with the `__export` one, which is deprecated).

scholzj commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,252 @@
+    # import resources
+    # KafkaTopic resources must be created *before*
+    # deploying the Topic Operator or it will delete them
+    oc apply -f $__TMP/resources 2>/dev/null ||true
```
This should use `kubectl`

scholzj commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,252 @@
+    done
+
+    # start statefulsets and operator
+    oc scale statefulset $CLUSTER_NAME-zookeeper --replicas $ZOO_REPLICAS
+    oc scale statefulset $CLUSTER_NAME-kafka --replicas $KAFKA_REPLICAS
```
You should not do this ... you should just scale the operator (or unpause the resource - see the other comment) and let the operator do this.

scholzj commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,252 @@
+    # start statefulsets and operator
+    oc scale statefulset $CLUSTER_NAME-zookeeper --replicas $ZOO_REPLICAS
+    oc scale statefulset $CLUSTER_NAME-kafka --replicas $KAFKA_REPLICAS
+    oc scale deployment strimzi-cluster-operator --replicas 1
```
This should use `kubectl`

scholzj commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,252 @@
+    done
+
+    # stop operator and statefulsets
+    oc scale deployment strimzi-cluster-operator --replicas 0
+    oc scale statefulset $CLUSTER_NAME-kafka --replicas 0
+    oc scale statefulset $CLUSTER_NAME-zookeeper --replicas 0
```
This should use `kubectl`

tombentley replied:
Specifically `kubectl wait`, I guess.

scholzj commented on docker-images/cold-backup/Makefile:
```diff
@@ -0,0 +1,9 @@
+PROJECT_NAME=cold-backup
```
If we really want to have this in this repository, it needs to be plugged into the build - probably by adding it here: https://github.com/strimzi/strimzi-kafka-operator/blob/37bbc7181abe3096daf8490c67e2912fa2d3d5f8/docker-images/build.sh#L12

scholzj reviewed (commented):
I left bunch of initial comments. One thing to note is that this should work with any Kubernetes and with `kubectl`. So it should either use `kubectl` directly or similarly to https://github.com/strimzi/strimzi-kafka-operator/blob/master/tools/report.sh it needs to detect which one is present and use the right one.

tombentley commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,252 @@
+    && { __HOME=$PWD; popd >/dev/null; }
+__MPOD="maintenance"
+
+SOURCE_NS="kafka"
+TARGET_NS="kafka"
+CLUSTER_NAME="my-cluster"
```
Perhaps these should be options to the command, or should be conditionally set, so you can at least pass them like this: `CLUSTER_NAME=different-cluster run.sh`.

fvaleri replied:
Good idea and I already considered that during development. Given that there are some of them, maybe a properties file would be better.

tombentley commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,252 @@
+# PVC size must match with the corresponding CR value
+# PVC class must be empty when not available
+# specify number of JBOD volumes or zero
+ZOO_REPLICAS="3"
```
For this and the `KAFKA_REPLICAS` can't we obtain the number from the CR using `kubectl get ... -o jsonpath ...` ?

fvaleri replied:
Sure. Will do that change.

tombentley commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,252 @@
+KAFKA_REPLICAS="3"
+KAFKA_PVC_SIZE="10Gi"
+KAFKA_PVC_CLASS=""
+KAFKA_JBOD="2"
```
Is this the number of volumes in the JBOD configuration? Maybe calling it a different name would be clearer?

fvaleri replied:
Agree. The name is not clear.

tombentley commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,252 @@
+        if [[ -n $crs ]]; then
+            echo "Exporting $name"
+            # delete runtime metadata expression
+            local exp="del(.metadata.namespace, .items[].metadata.namespace, \
+                .metadata.resourceVersion, .items[].metadata.resourceVersion, \
+                .metadata.selfLink, .items[].metadata.selfLink, \
+                .metadata.uid, .items[].metadata.uid, \
+                .status, .items[].status)"
```
I thought there was a `kubectl get --export` which removed this stuff?

fvaleri replied:
This flag is deprecated so I created this alternative function.

tombentley replied:
OK, that makes sense. But I wonder whether having an include list rather than an exclude list would be better, since we don't know what might get added in the future. But an include list might be tricker for `yq` to do.

tombentley commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,252 @@
+    # internal certificates and user secrets
+    __export "secrets" "strimzi.io/name=strimzi"
+    # custom configmap and secrets
+    for name in "${CUSTOM_CMS[@]}"; do
+        __export "cm/$name"
+    done
+    for name in "${CUSTOM_SECRETS[@]}"; do
+        __export "secret/$name"
+    done
```
I'd be tempted to just combine `CUSTOM_CMS` and `CUSTOM_SECRETS` into a single `CUSTOM_RESOURCES` which was a list of kind-prefixed names.

fvaleri replied:
Good idea. Thanks.

tombentley commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,252 @@
+}
+
+backup() {
+    __confirm $SOURCE_NS
```
We should probably make it clear, by printing a message, that backing up will cause downtime. Otherwise you're not really getting _informed_ consent.

fvaleri replied:
Agree. Thanks.

fvaleri commented:
@scholzj @tombentley Thanks for all suggestions, will work on them as soon as possible.

scholzj commented:
FYI: We discussed this on the Strimzi Community Meeting today. This should ultimately have its own repository in order to make sure the main repo doesn't grow too much and the CI times remain reasonable etc. But to make it easier we can finish the review here and move it only once it is ready. Thanks.

fvaleri commented:
@scholzj with the latest changes to make it `oc` free, I only need an image with `tar`, so I guess we can use the base image without impacts on CI time. Note that the backup is still incremental.

scholzj commented:
>  I only need an image with tar, so we can use the Strimzi base image without impacts on CI time. 

Ok, that would make things easier. Thanks for the update

fvaleri commented:
@tombentley @scholzj 2nd round of review after rebase and new commit. I should have addressed most of your suggestions and concerns. I'm now using `centos:7` image, instead of building a custom one (no impact on CI pipeline). I also added multi namespace/cluster support. Finally, the PR description is updated, not sure where is the most appropriate documentation place for it. 

Further improvements (separate PRs ?):
- Use `strimzi.io/pause-reconciliation` annotation (#4399) when version >= `0.22.0`, instead of shutting down the operator and potentially affecting any other cluster deployed in the same namespace. 
- Create a system test out of the manual test procedure. I don't have much experience with the Strimzi code base, but I can try to deal with it.

scholzj reviewed (commented):
Looks mostly good. I left some more comments. I think qe also need to have some clear description what it does and how it does it. I can for example imaging it as a README.md file in the `tools/cold-backup` or as a comment on the beginning of the script.

It is important that users understand what this does and what are the consequences of using it (outage).

scholzj commented:
> Create a system test out of the manual test procedure. I don't have much experience with the Strimzi code base, but I can try to deal with it.

I think that for example @im-konge should be able to guide you through it if needed.

im-konge commented:
@fvaleri Sure, I'm here if you'll need some help with the ST :)

tombentley commented:
More liberal comments in the script would not hurt, but otherwise I think this is looking good, assuming we can come up with the tests.

fvaleri commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,312 @@
+
+# custom configmap names
+# i.e. external logging configuration (log4j)
+CUSTOM_CMS=(
+    "log4j-properties"
+    "custom-test"
+)
```
Well, this list is there just an example for users. I agree that it doesn't make much sense now that we can pass parameters to the script. I will investigate how to dynamically get it from the CR and then add it to the `env` file.

fvaleri commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,312 @@
+
+# custom secret names
+# i.e. listener's certificate, registry authentication
+CUSTOM_SECRETS=(
+    "ext-listener-crt"
+    "registry-authn"
+    "custom-test"
+)
```
Same answer as above.

fvaleri commented:
> Looks mostly good. I left some more comments. I think qe also need to have some clear description what it does and how it does it. I can for example imaging it as a README.md file in the `tools/cold-backup` or as a comment on the beginning of the script.
> 
> It is important that users understand what this does and what are the consequences of using it (outage).

@scholzj Will add the README.md file with that description.

@tombentley Will add the ST and some more comments to facilitate maintenance.

fvaleri commented:
> @fvaleri Sure, I'm here if you'll need some help with the ST :)

Thanks @im-konge , I will probably need some.

fvaleri commented:
Added README and better input parameter parsing logic. Now will move on to implement the ST.

im-konge commented:
Hi @fvaleri - do you need any help with the STs?

fvaleri commented:
Hi @im-konge, I think I'm at a good point with the ST. I just need to find some time to add a couple of missing pieces and push it for a review. I will probably work on it during the weekend.

im-konge commented:
@fvaleri no rush, I was just wondering if you're not stuck on something :)

fvaleri commented:
Hi @im-konge , I rebased the branch and pushed the `ColdBackupScriptST` along with some minor script fixes/improvements. The ST seems to be robust and it takes around 10 minutes to complete using a remote K8s cluster running on AWS. I guess it should run a little faster on a local cluster. Please, let me know if there is something to improve.

To all: it would be good if you can give the bash script another try (see `tools/cold-backup/README.md` for the step-by-step test procedure).

Thanks.

im-konge commented on systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/KafkaClientOperations.java:
```diff
@@ -46,4 +48,12 @@
      * @throws Exception exception
      */
     int receiveMessagesTls(long timeoutMs) throws Exception;
+
+    /**
+     * Getting consumer group current offsets
+     * @param timeoutMs timeout in millisecondsß
+     * @return command output
+     * @throws Exception exception
+     */
+    Map<String, String> getCurrentOffsets(long timeoutMs) throws Exception;
```
Do you really need it here? In other classes where is the `KafkaClientOperations` class extended you are throwing the `throw new UnsupportedOperationException();` -> wouldn't it be easier to add it just to `InternalKafkaClient`?

fvaleri replied:
Agree.

im-konge reviewed (commented):
You've done a really good work with the ST, I have some suggestions/nits. Anyway - good job :)

im-konge commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,143 @@
+import io.strimzi.test.executor.Exec;
+import io.strimzi.test.executor.ExecResult;
+
+public class ColdBackupScriptST extends AbstractST {
```
Can you please add the `@Tag(REGRESSION)`?

fvaleri replied:
Sure.

fvaleri commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,143 @@
+
+    private static final Logger LOGGER = LogManager.getLogger(ColdBackupScriptST.class);
+    private static final String NAMESPACE = "cold-backup";
+    private static final String CLUSTER_NAME = "my-cluster";
+    private static final String TOPIC_NAME = "my-topic";
+    private static final String CLIENTS_NAME = CLUSTER_NAME + "-" + Constants.KAFKA_CLIENTS;
```
Sure. Thanks.

fvaleri commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,143 @@
+    @Test
+    @Tag(INTERNAL_CLIENTS_USED)
+    void backupAndRestore() {
+        int kafkaReplicas = 1, zooReplicas = 1;
```
I used them for logging purpose too. It works fine with 3, but I switched to 1 to speedup a bit.

fvaleri commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,143 @@
+
+        // deploy a test cluster
+        LOGGER.info("Deploying cluster {} with {} Kafka nodes and {} Zookeeper nodes", CLUSTER_NAME, kafkaReplicas, zooReplicas);
+        installClusterOperator(NAMESPACE);
```
Given that I'm reinstalling it before running the restore prcedure, I think it's better to handle it at the method stack level.

fvaleri commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,143 @@
+        assertEquals(batchSize, internalClients.receiveMessagesPlain());
+    }
+
+    private String deployInternalClients() {
```
Make sense. Thanks.

fvaleri commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,143 @@
+        return internalClientsPodName;
+    }
+
+    private InternalKafkaClient buildInternalClients(String podName, String groupName, int messageCount) {
+        InternalKafkaClient clients = new InternalKafkaClient.Builder()
+            .withListenerName(Constants.PLAIN_LISTENER_DEFAULT_NAME)
+            .withNamespaceName(NAMESPACE)
+            .withClusterName(CLUSTER_NAME)
+            .withTopicName(TOPIC_NAME)
+            .withUsingPodName(podName)
+            .withConsumerGroupName(groupName)
+            .withMessageCount(messageCount)
+            .build();
+        return clients;
+    }
```
Well, you can't set topicName anf groupName on InternalKafkaClient, that's why I'm doing this.

im-konge replied:
You can do
```
clients = clients.toBuilder()
            .withTopicName(TOPIC_NAME)
            .withConsumerGroupName(groupName)
            .build();
```
also, if you don't specify the `consumerGroup`, it will generate random one, so you don't need to bother with it.

fvaleri replied:
OK

fvaleri commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,143 @@
+        runExternalCommand(backupCmd);
+
+        // delete the namespace and restore
+        ResourceManager.deleteMethodResources();
```
Agree. Thanks.

fvaleri commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,143 @@
+
+        // deploy the operator and wait for provisionig
+        installClusterOperator(NAMESPACE);
+        waitForKafkaReady(CLUSTER_NAME);
```
Sorry, I don't get this.

fvaleri replied:
OK

fvaleri commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,143 @@
+        installClusterOperator(NAMESPACE);
+        waitForKafkaReady(CLUSTER_NAME);
+        internalClientsPodName = deployInternalClients();
+        internalClients = buildInternalClients(internalClientsPodName, groupName, secondBatchSize);
```
Nope, here I need to rebuild them, as I have a clean namespace where I just deployed the clients pod.

im-konge replied:
`internalClients = internalClients.toBuilder().withUsing..().build()` for example

fvaleri replied:
OK

im-konge commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,143 @@
+        // check consumer group offsets
+        LOGGER.info("Checking consumer group offsets after restore from pod: {}", internalClientsPodName);
+        Map<String, String> offsetsAfterRestore = internalClients.getCurrentOffsets();
+        logConsumerGroupOffsets(offsetsAfterRestore);
+        assertEquals(offsetsBeforeBackup, offsetsAfterRestore);
```
How I wrote, we don't want to log all offsets in `logConsumerGroupOffsets(offsetsAfterRestore);`, so the assert is still enough and this comment is still not resolved.

im-konge commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,144 @@
+
+        // deploy the operator and wait for provisionig
+        installClusterOperator(NAMESPACE);
+        waitForKafkaReady(clusterName);
```
```suggestion
        KafkaUtils.waitForKafkaReady(clusterName);
```

fvaleri replied:
OK

im-konge commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,143 @@
+        String[] backupCmd = new String[] {
+            PROJECT_HOME + "/tools/cold-backup/run.sh", "-by", "-n", NAMESPACE, "-c", CLUSTER_NAME, "-t", BACKUP_PATH
+        };
+        runExternalCommand(backupCmd);
```
@fvaleri still valid

fvaleri replied:
Correct, this came out of an experiment. Not needed anymore.

im-konge commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,143 @@
+    private static final String CLUSTER_NAME = "my-cluster";
+    private static final String TOPIC_NAME = "my-topic";
+    private static final String CLIENTS_NAME = CLUSTER_NAME + "-" + Constants.KAFKA_CLIENTS;
+    private static final String PROJECT_HOME = Paths.get("").toFile().getAbsolutePath();
```
@fvaleri still valid

fvaleri replied:
OK

fvaleri commented:
@im-konge , with latest commit, I should have addressed all your concerns/suggestions. Thanks!

fvaleri commented:
Just saw a couple of additional suggestions that I'm going to test.

scholzj commented on tools/cold-backup/README.md:
```diff
@@ -0,0 +1,85 @@
+./tools/cold-backup/run.sh -r -n myproject -c my-cluster -s /tmp/my-cluster.zip
+
+# deploy the operator and wait for provisionig
+kubectl create -f ./install/cluster-operator
```
Who and where installs the Kafka custom resource? I assume the backup/restore tool does it? You should then probably install the operator first before running the restore. In case you would be restoring on a different cluster, you might not have the CRDs installed until this step and that would fail the restore.

fvaleri replied:
Yes, I thought about this, but in practice doing like this ends with zookeeper error on TLS certificates. More details later.

fvaleri replied:
This is the issue with switching these two steps in the ST. It seem to work when doing it manually. Will do some more tests.

```
2021-03-16 15:26:05,574 WARN Exception caught (org.apache.zookeeper.server.NettyServerCnxnFactory) [nioEventLoopGroup-7-2]
io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: PKIX path validation failed: java.security.cert.CertPathValidatorException: signature check failed
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:471)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

fvaleri replied:
Ok I fixed that and switched the operator deploy before restore procedure.

scholzj commented on tools/cold-backup/README.md:
```diff
@@ -0,0 +1,85 @@
+backups, you should look at *MirrorMaker2* to sync with a remote cluster, but this comes with
+additional complexities and required resources.
+
+To run the script you must be logged in as *cluster-admin* user. The script will stop the operator
```
Does it really need to be _cluster-admin_ user? I don't see anything what would require it. It might be better to list the permissions the user needs (not necessarily listing every RBAC, but for example _you need to be abel to work with PVC and use the Strimzi custom resources_ etc.)

fvaleri replied:
Right. Let me double check this.

scholzj reviewed (commented):
Great to see the ST added. I left some more comments.

im-konge commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,123 @@
+
+    @Test
+    void backupAndRestore() {
+        String topicName = TOPIC_NAME;
```
I guess you don't need this, just use `TOPIC_NAME` directly

fvaleri replied:
Yep.

im-konge commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,123 @@
+
+        // check consumer group offsets
+        Map<String, String> offsetsAfterRestore = clients.getCurrentOffsets();
+        assertThat("Consumer group current offsets are same as before backup", offsetsAfterRestore, is(offsetsBeforeBackup));
+
+        // check consumer group recovery
+        assertThat("Consumer group is able to recover after restore", clients.receiveMessagesPlain(), is(secondBatchSize));
```
When you add comment in the `assertThat`, it shows when the assert fails, so you should use inverse logic.
For example, if the assert fails, you will get `Consumer group is able to recover after restore`, which is not correct.

fvaleri replied:
Ok. Thanks.

im-konge commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,123 @@
+            .withConsumerGroupName(newGroupName)
+            .withMessageCount(batchSize)
+            .build();
+        assertThat("A new consumer group is able to get all messages", clients.receiveMessagesPlain(), is(batchSize));
```
Same as above

fvaleri replied:
Ok. Thanks.

im-konge reviewed (commented):
Small nits (I guess last one), thanks for the ST!

fvaleri commented on tools/cold-backup/README.md:
```diff
@@ -0,0 +1,85 @@
+
+## Requirements
+- bash 5+ (GNU)
+- kubectl 1.20+ (Kubernetes)
```
Probably not. I was just using the one provided by `brew`. Let me do another test and change it to the supported one.

Frawless replied:
I guess if STs pass, we can just change the requirement to 1.16+

fvaleri replied:
Hi @Frawless , I just pushed a new commit tested with 1.16. I also added a 5s sleep before running the backup script, in order to make the test more robust.

fvaleri commented on .gitignore:
```diff
@@ -11,6 +11,7 @@ release.properties
 **/.prefs
 **/.classpath
 /target/
+**/bin/*
```
Ok, I'm going to remove it.

scholzj replied:
I have no idea why would VSCode create anything there - IntelliJ does not create there anything. But the folder contains the startup scripts. So it cannot be ignored.

scholzj replied:
I guess you either need to ignore the specific files and change VSCode to not add any files there.

fvaleri replied:
Ok, this is not a problem at all, but in these bin directories I see a mix of source and target files. Shouldn't we have something like the following in the gitignore, rather than having to create a manual configuration at the editor level?
```
# Generated assembly files
**/bin/**/*.class
**/bin/**/*.yaml
**/bin/**/*.out
**/bin/**/*.out
api-conversion/bin/src/*
api-conversion/bin/Makefile
api-conversion/bin/pom.xml
api-conversion/bin/README.md
```

scholzj replied:
TBH, I think you need to figure out what adds the file there. There is nothing like that in this directory in the master branch: https://github.com/strimzi/strimzi-kafka-operator/tree/master/api-conversion/bin and there is nothing like that in the bin directory in any of my working clones. The content you seem to have there suggests something just copied stuff there by mistake / error. So in general, the better way would be to make sure it is not there in the first place.

fvaleri replied:
Ok, I will revert that chaneg as it is unrelated to this PR. Anyway, I'm just using maven to do the project build, so it must be the output of some plugin.

Frawless commented:
/azp run regression

Frawless commented:
@fvaleri  it seems that your test failed on azure - https://dev.azure.com/cncf/strimzi/_build/results?buildId=69306&view=logs&j=cc43089d-b582-5629-f6a8-8a929a66ea7b&t=b295259a-8ce1-5ba9-7745-ccf7768a68b7  , did you already fix that or it's something it you didn't hit yet?

fvaleri commented:
> @fvaleri it seems that your test failed on azure - https://dev.azure.com/cncf/strimzi/_build/results?buildId=69306&view=logs&j=cc43089d-b582-5629-f6a8-8a929a66ea7b&t=b295259a-8ce1-5ba9-7745-ccf7768a68b7 , did you already fix that or it's something it you didn't hit yet?

I pushed a couple of fixes/improvements after you launched it. Can you retry?

Frawless commented:
@fvaleri sure

Frawless commented:
/azp run regression

scholzj commented:
Looks like the test failed:

```
[ERROR] Errors: 
[ERROR] io.strimzi.systemtest.backup.ColdBackupScriptST.backupAndRestore
[ERROR]   Run 1: ColdBackupScriptST.backupAndRestore:90 » NullPointer
[ERROR]   Run 2: ColdBackupScriptST.backupAndRestore:90 » NullPointer
[ERROR]   Run 3: ColdBackupScriptST.backupAndRestore:90 » NullPointer
```

fvaleri commented:
> Looks like the test failed:
> 
> ```
> [ERROR] Errors: 
> [ERROR] io.strimzi.systemtest.backup.ColdBackupScriptST.backupAndRestore
> [ERROR]   Run 1: ColdBackupScriptST.backupAndRestore:90 » NullPointer
> [ERROR]   Run 2: ColdBackupScriptST.backupAndRestore:90 » NullPointer
> [ERROR]   Run 3: ColdBackupScriptST.backupAndRestore:90 » NullPointer
> ```

Ok let me look into it.

im-konge commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,125 @@
+        // send additional messages
+        clients.setMessageCount(secondBatchSize);
+        clients.sendMessagesPlain();
+        TestUtils.sleep(5_000L);
```
Why you need this?

fvaleri replied:
I'm already working on removing it :)

fvaleri commented:
@scholzj @im-konge @Frawless 

Ok, I fixed the script issue responsible for the test flackiness, removed the `sleep()` from the ST and also fixed the NPE on `waitForKafkaStatus()`. I also reverted back the `.gitignore` change. Now the ST is consistenly green, both using minikube and a remote OCP cluster. It takes under 10 minutes with all images cached.

Can you run the Azure test pipeline one more time?

Frawless commented:
/azp run regression

im-konge commented on systemtest/src/main/java/io/strimzi/systemtest/utils/kafkaUtils/KafkaUtils.java:
```diff
@@ -68,7 +68,11 @@ public static void waitForKafkaNotReady(String clusterName) {
 
     public static void waitForKafkaStatus(String clusterName, Enum<?>  state) {
-        Kafka kafka = KafkaResource.kafkaClient().inNamespace(kubeClient().getNamespace()).withName(clusterName).get();
+        String namespace = kubeClient().getNamespace();
+        Kafka kafka = KafkaResource.kafkaClient().inNamespace(namespace).withName(clusterName).get();
+        if (kafka == null) {
+            throw new RuntimeException(String.format("Kafka resource %s not found in namespace %s", clusterName, namespace));
+        }
```
If `waitForKafkaStatus` throwed NPE, this will throw exception on the same place and the test will end here.
Maybe it would be great to wait for the Kafka (re)creation and then `waitForKafkaStatus`.

Something like this:
```
    public static void waitForKafkaRecreation(String clusterName) {
        LOGGER.info("Waiting for recreation of Kafka:{}", clusterName);
        TestUtils.waitFor("Kafka recreation " + clusterName, Constants.POLL_INTERVAL_FOR_RESOURCE_READINESS, Constants.GLOBAL_TIMEOUT,
            () -> KafkaResource.kafkaClient().inNamespace(kubeClient().getNamespace()).withName(clusterName).get() != null);

        LOGGER.info("Kafka: {} recreated.", clusterName);
    }
```
and you can add it to `KafkaUtils`.
Then you'll do just
```
KafkaUtils.waitForKafkaRecreation(clusterName);
KafkaUtils.waitForKafkaReady(clusterName);
```

fvaleri replied:
The ST is still failing at the same point on Azure, this time is captured by my `RuntimeException`. You may be right, but I'm wondering why this never happens on my local env. Anyway, thanks! I'll review later.

### Mine
```log
2021-03-17 15:58:38 INFO  [Exec:172] ======STDOUT END======
2021-03-17 15:58:38 INFO  [ColdBackupScriptST:88] Waiting for provisioning of restored cold-backup/my-cluster-1300602412
2021-03-17 15:58:38 DEBUG [CustomResource:157] Calling CustomResource#setKind doesn't do anything because the Kind is computed and shouldn't be changed
2021-03-17 15:58:38 INFO  [ResourceManager:449] Wait for Kafka: my-cluster-1300602412 will have desired state: Ready
...
2021-03-17 15:59:05 INFO  [TimeMeasuringSystem:113] End time of operation TEST_EXECUTION is correctly stored
2021-03-17 15:59:05 INFO  [AbstractST:655] Search in strimzi-cluster-operator log for errors in last 442 seconds
2021-03-17 15:59:06 INFO  [BaseCmdKubeClient:439] Exception not found
```

### Azure
```log
2021-03-17 16:35:47 INFO  [Exec:172] ======STDOUT END======
2021-03-17 16:35:47 INFO  [ColdBackupScriptST:88] Waiting for provisioning of restored cold-backup/my-cluster-900682292
2021-03-17 16:35:48 INFO  [TimeMeasuringSystem:113] End time of operation TEST_EXECUTION is correctly stored
...
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 745.473 s <<< FAILURE! - in io.strimzi.systemtest.backup.ColdBackupScriptST
[ERROR] io.strimzi.systemtest.backup.ColdBackupScriptST.backupAndRestore  Time elapsed: 733.172 s  <<< ERROR!
java.lang.RuntimeException: Kafka resource my-cluster-900682292 not found in namespace cold-backup
	at io.strimzi.systemtest.utils.kafkaUtils.KafkaUtils.waitForKafkaStatus(KafkaUtils.java:74)
	at io.strimzi.systemtest.utils.kafkaUtils.KafkaUtils.waitForKafkaReady(KafkaUtils.java:63)
	at io.strimzi.systemtest.backup.ColdBackupScriptST.backupAndRestore(ColdBackupScriptST.java:89)
```

im-konge replied:
The problem is, that on the other env it can be somehow slower than on your machine, so these things are causing race conditions. When you add the wait there and revert changes in `waitForKafkaStatus`, it will be okay.

fvaleri replied:
@im-konge I'm not able to reproduce the issue on my end probably because of different timing. 

Just committed your suggestion, let's see if it works now.

im-konge replied:
Okay, if it will work, revert please changes in `waitForKafkaStatus`. Thanks! :)

fvaleri replied:
ok

im-konge replied:
So the test failed again, because Kafka is not present in 5 minutes. Is it possible that is required k8s version `1.20.0`+? On Azure we are running tests on k8s `1.16.0`, so I triggered Jenkins pipeline (where is `1.20+`) to see if it will have any effect.

fvaleri replied:
Ok, let me know. Have you tried to run it locally?

Frawless commented:
/azp run regression

im-konge commented:
@strimzi-ci run tests test_only profile=regression testcase=ColdBackupScriptST

fvaleri commented:
> @strimzi-ci run tests test_only profile=regression testcase=ColdBackupScriptST

Ok, let me know. Have you tried to run it locally?

im-konge commented:
@strimzi-ci run tests test_only profile=regression testcase=ColdBackupScriptST

im-konge commented:
@strimzi-ci run tests test_only profile=regression testcase=ColdBackupScriptST

im-konge commented:
@strimzi-ci run tests test_only profile=regression testcase=ColdBackupScriptST

im-konge commented:
@strimzi-ci run tests test_only profile=regression testcase=ColdBackupScriptST

im-konge commented:
@strimzi-ci run tests test_only profile=regression testcase=ColdBackupScriptST

fvaleri commented:
hi @im-konge, reverted KafkaUtils changes.

Since the previous commit, the ST always works fine on my local and remote clusters. Have you tried running it locally? 

It may be an environment issue. How can I get logs from Azure pods to troubleshoot further?

im-konge commented:
Hi @fvaleri , I started build on my minikube with version `1.16.0` and then I'll try that again on `1.20.0`, in the meantime I tried your test on OCP4.7, where it was successful. So give please some time to test it and I'll let you know.

And about logs from Azure - https://dev.azure.com/cncf/strimzi/_build/results?buildId=69464&view=artifacts&pathAsName=false&type=publishedArtifacts
here is yours.

Thanks for your patience :)

fvaleri commented:
> Hi @fvaleri , I started build on my minikube with version `1.16.0` and then I'll try that again on `1.20.0`, in the meantime I tried your test on OCP4.7, where it was successful. So give please some time to test it and I'll let you know.
> 
> And about logs from Azure - https://dev.azure.com/cncf/strimzi/_build/results?buildId=69464&view=artifacts&pathAsName=false&type=publishedArtifacts
> here is yours.
> 
> Thanks for your patience :)

Sure, no problem. Take your time and thanks for the help :)

fvaleri commented:
Found the issue: for some reason the `waitForKafkaRecreation` runs on another namespace on Azure, probably `default`. You can fix it by adding the `namespace` as an additinal input parameter instead of calling`kubeClient().getNamespace()`. In the end, we don't actually need it anymore as the restore procedure now waits for cluster recreation.

I will commit the fix in a bit, this time the ST should work fine.

scholzj commented:
/azp run regression

im-konge commented:
@scholzj @fvaleri so I've done some tests:
on k8s `v1.16.0` test is constantly failing
on k8s `v1.20.0` test is always passing

So I guess that we need k8s `v1.20.0`+ to running this.

fvaleri commented:
> @scholzj @fvaleri so I've done some tests:
> on k8s `v1.16.0` test is constantly failing
> on k8s `v1.20.0` test is always passing
> 
> So I guess that we need k8s `v1.20.0`+ to running this.

Is this with my latest commit?

im-konge commented:
@fvaleri Nope, it was with previous one, let me check it again.

scholzj commented:
Just FYI: I do not think that if it works only on 1.20+ it would be a blocker (although it would be good to understand why since I do not see anything what would be Kube version specific). We just need to describe it in the README and gate the ST to run only on 1.20+. But otherwise it would be fine.

fvaleri commented:
I can confirm that the bash script works on OCP 3.11 (K8s 1.11), so it's just a matter of fixing the issue with the ST. I will get a 1.16 env and look at it in the next few days.

fvaleri commented:
I cannot be 100% sure b/c I have no NPE with K8s 1.16 on my machine, but I suspect that the Azure pipeline issue comes from clients pod not being properly initialized (wrong ns is selected). I commited a new a version that rebuilds clients from scratch after restore.

scholzj commented:
/azp run regression

see-quick commented:
Hi, just a note. Recently, we merged a huge PR (https://github.com/strimzi/strimzi-kafka-operator/pull/4137), relating to `system test module`. Almost everything is changed. So, when you rebase you will have all changes which bind to our refactor. If you have some problems with re-writing the tests I am here to help.

scholzj commented:
@see-quick Maybe you can provide a patch?

fvaleri commented:
@im-konge @scholzj not sure I will be able to work on it today, but we have a ns warning on Azure right after the backup, when we delete and recreate the ns.

```log
# azure (NPE)
2021-03-20 20:47:00 INFO  [NamespaceUtils:23] Waiting for Namespace cold-backup deletion
2021-03-20 20:47:44 INFO  [NamespaceUtils:27] Namespace cold-backup was deleted
2021-03-20 20:47:44 INFO  [AbstractST:137] Going to install ClusterOperator via Yaml bundle
2021-03-20 20:47:44 WARN  [KubeClusterResource:149] Namespace cold-backup is already created, going to delete it
2021-03-20 20:47:50 INFO  [KubeClusterResource:154] Creating Namespace cold-backup
2021-03-20 20:47:50 INFO  [KubeClusterResource:160] Using Namespace cold-backup
2021-03-20 20:47:50 INFO  [KubeClusterResource:80] Changing to cold-backup namespace

# mine (success)
2021-03-20 18:40:52 INFO  [NamespaceUtils:23] Waiting for Namespace cold-backup deletion
2021-03-20 18:41:36 INFO  [NamespaceUtils:27] Namespace cold-backup was deleted
2021-03-20 18:41:36 INFO  [AbstractST:137] Going to install ClusterOperator via Yaml bundle
2021-03-20 18:41:36 INFO  [KubeClusterResource:154] Creating Namespace cold-backup
2021-03-20 18:41:36 INFO  [KubeClusterResource:160] Using Namespace cold-backup
2021-03-20 18:41:36 INFO  [KubeClusterResource:80] Changing to cold-backup namespace
```

im-konge commented:
The problem is, that you are creating namespace with `ResourceManager.kubeClient().createNamespace(NAMESPACE);` and then you are creating it with `installClusterOperator(NAMESPACE);` -> this method contains: if namespace already exists, delete it and create new one, apply CRBs etc., deploy operator. So, I think you can simply remove the `ResourceManager.kubeClient().createNamespace(NAMESPACE);` and keep just the `installClusterOperator(NAMESPACE);`

im-konge commented:
@fvaleri also, how @see-quick mentioned, PR with huge refactor of STs is merged, so if you'll give me permissions to your repo, I can create branch with changes and then you can merge it, WDYT?

fvaleri commented:
> The problem is, that you are creating namespace with `ResourceManager.kubeClient().createNamespace(NAMESPACE);` and then you are creating it with `installClusterOperator(NAMESPACE);` -> this method contains: if namespace already exists, delete it and create new one, apply CRBs etc., deploy operator. So, I think you can simply remove the `ResourceManager.kubeClient().createNamespace(NAMESPACE);` and keep just the `installClusterOperator(NAMESPACE);`

Kind of make sense, but why I'm not seeing it on my local env running the same K8s version? Is it a matter of timing? Do you think this is related to the NPE we get after restore when using clients for the assertions?

> @fvaleri also, how @see-quick mentioned, PR with huge refactor of STs is merged, so if you'll give me permissions to your repo, I can create branch with changes and then you can merge it, WDYT?

Sure, but first we should make the ST work on Azure. I can't work on it today, but tomorrow I'll have a deeper look and wil give you permissions on my repo. Thanks.

scholzj commented:
I gave it a try and have some observations + probably the reason why it does not work:

1) It does not seem to work with JBOD storage. I think this is a significant limitation given we have it in all examples and recommend it normally even if users use a single disk.

2) It does not seem to work on MacOS because of sed, tar or readlink being used. I think it would be good to use https://github.com/strimzi/strimzi-kafka-operator/blob/master/tools/multi-platform-support.sh to get it working on MacOS as well similarly to our other tools. But maybe we can just add a note to the README.md and leave it for the next PR to not complicate this further.

3) The Kafka Exporter is not scaled down which sends it crash-looping. I think we can live with it, but it might trigger all kind of alarms in the user environment. So maybe you can consider scaling it down as well?

4) The tool is not restartable. Ig it fails, you have to manually clean-up afterwards to be able to run it again (scale up the CO and wait for it to scale up all pods, manually delete the strimzi-backup pod etc.). I think we can live with it, but if it was able to be restarted more smoothly it would be more user-friendly.

5) It looks like the tempoaryr structure in `/tmp/strimzi-backup` is not deleted at the end. It would be nice if it cleans up.

6) I run the tool like this `tools/cold-backup/run.sh -by -n myproject -c my-cluster -t ./backup.zip` from the root of my clone of the Strimzi repository. My expectation was that the `backup.zip` will be created there since that is what I specified. But it was instead created in `/tmp/strimzi-backup` which seems very unintuitive.

7) Now, this is the part which IMHO fails the test on Azure. When restoring the backup using `kubectl` from Kube 1.16, it does not restore the Kafka custom resource, so the Kafka CR is missing. The NPE is just a side-effect but not the real issue. The reason why it does not work is that the YAML file with the Kafka CR seems to be invalid. It looks like this:

```
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
# ...
items: []
```

where the `items` array makes it invalid and the `kubectl` from Kube 1.16 command fails with:

```
error: error validating "resources/myproject-kafkas.yaml": error validating data: ValidationError(Kafka): unknown field "items" in io.strimzi.kafka.v1beta2.Kafka; if you choose to ignore these errors, turn validation off with --validate=false
```

I assume this is does not happen with `kubectl` from Kube 1.16 and that is why it works there. But it looks like the way you expert the resource is wrong.

It also points to another issues which help this to happen:
* The restoring of the resources seems to hide its execution status with `||true` which makes it not fail
* The error messages is actually sent to `/dev/null` which seems wrong as well (+ ideally the backup / restore logs in the ST should be printed to STDOUT or stored in artifacts I guess)
* When the wait for the Entity Operator to be recreated ends without the operator running, it still reports success instead of error.
* The ST does not check Kafka readiness and just assumes it is running => this is what leads to the confusing NPE.

see-quick commented:
Here is the `patch` you have to apply. https://gist.github.com/see-quick/5eb0287a6375f61823298b76067406a1

see-quick commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,434 @@
+        echo "Rsync from $source to $target"
+        local pod_name="strimzi-backup"
+        local patch=$(sed "s/\$pvc/$pvc/g" $__HOME/patch.json)
+        kubectl run $pod_name --image="dummy" --restart="Never" --overrides="$patch"
```
What is the purpose of the dummy image here? Our tests are designed that CO will always try to download that image. You have to somehow modify `imagepullpolicy`...

fvaleri replied:
Not sure what you mean. That image is needed to spin up a maintenance pod that is at the heart of the script.

see-quick reviewed (commented):
Anyway, I tried your test but it seems to fail on that commend which I mentioned above.

fvaleri commented:
Hi @scholzj 

> I gave it a try and have some observations + probably the reason why it does not work:
> 
>     1. It does not seem to work with JBOD storage. I think this is a significant limitation given we have it in all examples and recommend it normally even if users use a single disk.

It should work, I included JBOD support right from the start as I know we recommend it. Do you have any specific error?

> 
>     2. It does not seem to work on MacOS because of sed, tar or readlink being used. I think it would be good to use https://github.com/strimzi/strimzi-kafka-operator/blob/master/tools/multi-platform-support.sh to get it working on MacOS as well similarly to our other tools. But maybe we can just add a note to the README.md and leave it for the next PR to not complicate this further.
> 

In the readme file, I specify that the GNU versions of these tools are required. That script you are referring to just creates aliases and doesn't check if they are actually installed, so it doesn't make it work on MacOS. 

I prefer to have these tools on my `$PATH` under the original name (without the `g` prefix). This is the way to do it by adding this simple snippet to your `~/.bash_profile`:

```sh
for i in /usr/local/Cellar/*/*/libexec/gnubin; do
  export PATH="$i:$PATH"
done
for i in /usr/local/Cellar/*/*/libexec/gnuman; do
  export MANPATH="$i:$MANPATH"
done
```

If you think it's useful, we may add this configuration and the exact `brew install` command in the readme.

>     3. The Kafka Exporter is not scaled down which sends it crash-looping. I think we can live with it, but it might trigger all kind of alarms in the user environment. So maybe you can consider scaling it down as well?
> 

I guess the alarms would fire up in any case, given that the cold backup procedure shuts down the entire cluster. Anyway this is not a problem and I can easily do this.

>     4. The tool is not restartable. Ig it fails, you have to manually clean-up afterwards to be able to run it again (scale up the CO and wait for it to scale up all pods, manually delete the strimzi-backup pod etc.). I think we can live with it, but if it was able to be restarted more smoothly it would be more user-friendly.
> 

Good point. I will add the on exit logic to do the cleanup.

>     5. It looks like the tempoaryr structure in `/tmp/strimzi-backup` is not deleted at the end. It would be nice if it cleans up.

Yeah, in theory `tar` supports also incremental backups. So we can do this only if this flag is NOT active (full backup).

> 
>     6. I run the tool like this `tools/cold-backup/run.sh -by -n myproject -c my-cluster -t ./backup.zip` from the root of my clone of the Strimzi repository. My expectation was that the `backup.zip` will be created there since that is what I specified. But it was instead created in `/tmp/strimzi-backup` which seems very unintuitive.
> 

Mmm, it should work as you expected. I'll double check this.

>     7. Now, this is the part which IMHO fails the test on Azure. When restoring the backup using `kubectl` from Kube 1.16, it does not restore the Kafka custom resource, so the Kafka CR is missing. The NPE is just a side-effect but not the real issue. The reason why it does not work is that the YAML file with the Kafka CR seems to be invalid. It looks like this:
> 
> 
> ```
> apiVersion: kafka.strimzi.io/v1beta2
> kind: Kafka
> # ...
> items: []
> ```
> 
> where the `items` array makes it invalid and the `kubectl` from Kube 1.16 command fails with:
> 
> ```
> error: error validating "resources/myproject-kafkas.yaml": error validating data: ValidationError(Kafka): unknown field "items" in io.strimzi.kafka.v1beta2.Kafka; if you choose to ignore these errors, turn validation off with --validate=false
> ```
> 
> I assume this is does not happen with `kubectl` from Kube 1.16 and that is why it works there. But it looks like the way you expert the resource is wrong.
> 
> It also points to another issues which help this to happen:
> 
>     * The restoring of the resources seems to hide its execution status with `||true` which makes it not fail
> 
>     * The error messages is actually sent to `/dev/null` which seems wrong as well (+ ideally the backup / restore logs in the ST should be printed to STDOUT or stored in artifacts I guess)
> 
>     * When the wait for the Entity Operator to be recreated ends without the operator running, it still reports success instead of error.
> 
>     * The ST does not check Kafka readiness and just assumes it is running => this is what leads to the confusing NPE.

Inside the backup resources folder, there should be only YAML files of CR that were existing in the source namespace. Inside a single YAML, you could have a list only when there are multiple resources of that kind (i.e. KafkaTopics). I'll investigate more on this as I'm still not able to reproduce it.

scholzj commented:
> It should work, I included JBOD support right from the start as I know we recommend it. Do you have any specific error?

I do not have the exact error at hand. But basically, it tired to find the PVC with the name corresponding to regular storage instead of the JBOD storage and failed.

In the readme file, I specify that the GNU versions of these tools are required. That script you are referring to just creates aliases and doesn't check if they are actually installed, so it doesn't make it work on MacOS.

> I prefer to have these tools on my $PATH under the original name (without the g prefix). This is the way to do it by adding this simple snippet to your ~/.bash_profile:
>
> ```
> for i in /usr/local/Cellar/*/*/libexec/gnubin; do
>   export PATH="$i:$PATH"
> done
> for i in /usr/local/Cellar/*/*/libexec/gnuman; do
>   export MANPATH="$i:$MANPATH"
> done
> ```
>
> If you think it's useful, we may add this configuration and the exact brew install command in the readme.

Well, whole Strimzi and its build system is using the `g`-prefix names of the utilities (https://github.com/strimzi/strimzi-kafka-operator/blob/master/development-docs/DEV_GUIDE.md#mac-os). So I think it is a bit weird to have suddenly this tool use something else. I think it is weird to tell users who have the correct setup to build Strimzi that they bow need to use something else to actually run the tools. And if we change this later, it might cause chaos because it just breaks the tool for users who used it before.

> Inside the backup resources folder, there should be only YAML files of CR that were existing in the source namespace. Inside a single YAML, you could have a list only when there are multiple resources of that kind (i.e. KafkaTopics). I'll investigate more on this as I'm still not able to reproduce it.

I have 3 files there. The KafkaTopic list looks ok as a valid list. The same for the list of secrets. Just the Kafka CR is a bit weird. I can share the whole file if needed - I still have the ZIP.

BTW: I think it needs to be more careful then just dump everything in the namespace sicne there are users with multiple clusters in one namespace. So it might be best to use the labels to dup the right secrets / topics / users and only the single Kafka CR.

fvaleri commented:
Just an update about waht I'm working on based on previous discussions:

- check JBOD storage type identification logic
- make it MacOS compatible (see `tools/multi-platform-support.sh`)
- add on exit cleanup (remove backup pod if present, clean temp dir, restart the cluster) **DONE**
- scale down to 0 also kafka exporter and entity operator **DONE**
- check the `-t` flag using `./backup.zip` as target path **DONE**
- check export resource logic (use label with cluster name)
- remove `||true` from restore procedure
- raise error if wait for entity operator ends with pod not running  **DONE**
- add Kafka readiness check to the ST 

When this is committed and reviewed, maybe we can have have another run of the ST.

fvaleri commented:
Ok, I fixed all previous points and did a fair amount of tests on K8s 1.16. 
Would you mind to run the Azure regression pipeline again? 
Thanks.

scholzj commented:
/azp run regression

scholzj commented:
@fvaleri I think you need to rebase it / fix the conflict first before I can run the regression.

fvaleri commented:
ok, I'll do it.

scholzj commented:
/azp run regression

fvaleri commented:
It failed with:
```log
error: error validating "/tmp/strimzi-backup/cold-backup/my-cluster-1069257492/resources/kafka-my-cluster-1069257492.yaml": error validating data: ValidationError(Kafka): unknown field "items" in io.strimzi.kafka.v1beta2.Kafka; if you choose to ignore these errors, turn validation off with --validate=false
```

I have changed the `__export_res` function like this, where type is "kafka" and id is "$CLUSTER_NAME". 
```sh
local resources=$(kubectl get $type -l $id -o name --ignore-not-found)
if [[ -z $resources ]]; then
    resources=$(kubectl get $type $id -o name --ignore-not-found 2>/dev/null ||true)
fi
```

And, in my local environment I'm not getting a list of Kafka anymore, but just the target Kafka cluster:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  creationTimestamp: "2021-03-29T05:50:41Z"
  generation: 1
  name: my-cluster
spec:
  # ...
```

It seems that on Azure these lines are still producing an empty result (`items []`). I will investigate more on this and also try to speedup the cluster start at the end of the restore procedure.

scholzj commented on tools/multi-platform-support.sh:
```diff
@@ -24,4 +25,5 @@ then
     WC=gwc
     UNIQ=guniq
     SORT=gsort
+    TAR=gtar
```
You will need to add also `readlink` / `greadlink` here.

scholzj commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,482 @@
+set -Eeuo pipefail
+#set -x #debug
+__TMP="/tmp/strimzi-backup" && mkdir -p $__TMP
+__HOME="" && pushd "$(dirname "$(readlink -f "${BASH_SOURCE[0]}")")" >/dev/null \
+    && { __HOME=$PWD; popd >/dev/null; }
```
This does not seem to work for me ... when I run it from any other directory then directly whre the `run.sh` is, it fails to load the multiple platform support.

fvaleri commented:
OK, I was able to reproduce the issue. This commit should fix the issue upgrading `yq` to v4.6.3.

scholzj commented:
/azp run regression

fvaleri commented:
@scholzj did we hit this one?

https://github.com/kubernetes/minikube/issues/10859

scholzj commented:
Looks like it. But I do not think I understand what does the error mean. Does that com from your handling of the context?

fvaleri commented:
Commented a MacOS compatibility snippet that was updating the path. Let's try without this.

scholzj commented:
/azp run regression

fvaleri commented:
For security reasons, I will commit a little script change to add readonly for all variables that are used with `rm -rf` command.

scholzj commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,481 @@
+        ZK_PVC_SIZE=$(kubectl -n $NAMESPACE get pvc $filter-0 -o yaml | yq eval ".spec.resources.requests.storage" -)
+        ZK_PVC_CLASS=$(kubectl -n $NAMESPACE get pvc $filter-0 -o yaml | yq eval ".spec.storageClassName" -)
+        local kafka_storage_type=$(kubectl -n $NAMESPACE get kafka $cluster_name -o yaml | yq eval ".spec.kafka.storage.type" -)
+        if [[ $kafka_storage_type == "jbod" ]]; then
+            JBOD_VOL_NUM=$(kubectl -n $NAMESPACE get pvc | grep $cluster_name-kafka-0 | wc -l)
+            filter="data-0-$cluster_name-kafka"
+        else
```
How does this handle when the JBOD storage is using more then one volume? Also, keep in mind that the volume IDs are customizable. So users might for example have JBOD volumes with IDs 3, 5 and 8. (I did not tried JBOD with multiple volumes, so maybe it is handled fine ... this just made me wonder if it does)

fvaleri replied:
OK

scholzj commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,481 @@
+        __rsync $pvc $path
+    done
+    for ((i = 0; i < $KAFKA_REPLICAS; i++)); do
+        if ((JBOD_VOL_NUM > 1)); then
```
What if I have JBOD storage with only one volume? This is probably why it does not work for me.

fvaleri replied:
OK

scholzj reviewed (commented):
I left some more minor comments. But it also did not worked for me in my environment with JBOD storage - I left some comments on the places I thought are the problem. If needed, full debug log is here: https://gist.github.com/scholzj/ace43f3f5e542a72a07b884bba927681

PaulRMellor commented on tools/cold-backup/README.md:
```diff
@@ -0,0 +1,82 @@
+Bash script for cold/offline backups of Kafka clusters on Kubernetes/OpenShift.
+
+If you think you do not need a backup strategy for Kafka as it has embedded data replication,
+then try to immagine a misconfiguration/bug/security-breach deleting all your data. For hot/online
```
```suggestion
then consider the impact of a bug or security breach that deleting all your data. For hot/online
```

fvaleri replied:
Thanks :)

PaulRMellor commented on tools/cold-backup/README.md:
```diff
@@ -0,0 +1,82 @@
+
+If you think you do not need a backup strategy for Kafka as it has embedded data replication,
+then try to immagine a misconfiguration/bug/security-breach deleting all your data. For hot/online
+backups, you should probably look at storage snapshotting or streaming into object storage.
```
```suggestion
backups, you can use storage snapshotting or streaming into object storage.
```

fvaleri replied:
Thanks :)

PaulRMellor commented on tools/cold-backup/README.md:
```diff
@@ -0,0 +1,82 @@
+then try to immagine a misconfiguration/bug/security-breach deleting all your data. For hot/online
+backups, you should probably look at storage snapshotting or streaming into object storage.
+
+To run the script, the K8s user must have rights to work with PVC and use Strimzi custom resources.
```
```suggestion
To run the script, the Kubernetes user must have permission to work with PVC and Strimzi custom resources.
```

fvaleri replied:
Thanks :)

PaulRMellor commented on tools/cold-backup/README.md:
```diff
@@ -0,0 +1,82 @@
+backups, you should probably look at storage snapshotting or streaming into object storage.
+
+To run the script, the K8s user must have rights to work with PVC and use Strimzi custom resources.
+The backup procedure will stop operator and selected cluster for the entire duration. Before restoring
+the Kafka cluster you need to make sure to have the right version of Strimzi CRDs installed. If you
+have a single cluster-wide operator, then you need to scale it down manually. You can run backup and
+restore procedures of different Kafka clusters in parallel. Only local file system is supported.
+ConsumerGroup offsets are included, but not KafkaConnect, MirrorMaker and KafkaBridge CRs.
```
```suggestion
The procedure will stop the Cluster Operator and selected cluster for the duration of the backup. Before restoring
the Kafka cluster you need to make sure to have the right version of Strimzi CRDs installed. If you
have a single cluster-wide Cluster Operator, then you need to scale it down manually. You can run backup and
restore procedures for different Kafka clusters in parallel. Only the local file system is supported.
`ConsumerGroup` offsets are included, but not Kafka Connect, MirrorMaker and Kafka Bridge custom resources.
```

fvaleri replied:
Thanks :)

PaulRMellor commented on systemtest/src/test/java/io/strimzi/systemtest/backup/ColdBackupScriptST.java:
```diff
@@ -0,0 +1,121 @@
+        clientsPodName = deployAndGetInternalClientsPodName(context);
+        clients = buildInternalClients(context, clientsPodName, groupName, secondBatchSize);
+        Map<String, String> offsetsAfterRestore = clients.getCurrentOffsets();
+        assertThat("Consumer group current offsets are not the same as before backup", offsetsAfterRestore, is(offsetsBeforeBackup));
```
```suggestion
        assertThat("Current consumer group offsets are not the same as before the backup", offsetsAfterRestore, is(offsetsBeforeBackup));
```

fvaleri replied:
Thanks :)

PaulRMellor commented on tools/cold-backup/README.md:
```diff
@@ -0,0 +1,82 @@
+# Strimzi backup
+Bash script for cold/offline backups of Kafka clusters on Kubernetes/OpenShift.
+
+If you think you do not need a backup strategy for Kafka as it has embedded data replication,
```
```suggestion
If you think you do not need a backup strategy for Kafka because of its embedded data replication,
```

fvaleri replied:
Thanks :)

PaulRMellor reviewed (approved):
Few suggestions

fvaleri commented:
@PaulRMellor thanks, they all make sense :) 

@scholzj I'm rewriting the `__export_env` function to be more resilient and work with any storage config.

fvaleri commented on tools/cold-backup/run.sh:
```diff
@@ -0,0 +1,481 @@
+#!/usr/bin/env bash
+set -Eeuo pipefail
+#set -x #debug
```
Removed.

fvaleri commented on tools/cold-backup/README.md:
```diff
@@ -0,0 +1,82 @@
+Bash script for cold/offline backups of Kafka clusters on Kubernetes/OpenShift.
+
+If you think you do not need a backup strategy for Kafka as it has embedded data replication,
+then try to immagine a misconfiguration/bug/security-breach deleting all your data. For hot/online
```
OK

fvaleri commented:
Ok, now it should be fine. I also improved the user interface (commands/options):
```sh
./run.sh backup -n $NAMESPACE -c my-cluster -t /tmp/my-cluster.zip -m kafka-metrics
```

@scholzj, when you have time, give it another try. Thanks.

scholzj commented:
/azp run regression

scholzj commented:
@fvaleri Seems to work fine for me locally now. I started the regression tests to have that run as well.

@Frawless @im-konge @see-quick @tombentley Could you have another look at this?

im-konge reviewed (approved):
I was thinking if the change of `boolean` to `void` in `waitForKafkaStatus` and `waitForKafkaReady` will cause some issues, but I don't see any usage of the `boolean`, so I guess that's good. LGTM, thanks for writing the ST :)

scholzj commented:
Thanks for the PR @fvaleri
